db/external_sst_file_test.cc

```diff
@@ -1700,6 +1700,43 @@ TEST_F(ExternalSSTFileTest, SstFileWriterNonSharedKeys) {
   ASSERT_OK(DeprecatedAddFile({file_path}));
 }
 
+TEST_F(ExternalSSTFileTest, WithUnorderedWrite) {
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->LoadDependency(
+      {{"DBImpl::WriteImpl:UnorderedWriteAfterWriteWAL",
+        "ExternalSSTFileTest::WithUnorderedWrite:WaitWriteWAL"},
+       {"ExternalSSTFileTest::WithUnorderedWrite:ContinueWriteMemtable",
+        "DBImpl::WriteImpl:BeforeUnorderedWriteMemtable"}});
+  SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::IngestExternalFile:NeedFlush", [&](void* need_flush) {
+        ASSERT_TRUE(*reinterpret_cast<bool*>(need_flush));
+      });
+
+  Options options = CurrentOptions();
+  options.unordered_write = true;
+  DestroyAndReopen(options);
+  Put("foo", "v1");
+  SyncPoint::GetInstance()->EnableProcessing();
+  port::Thread writer1([&]() { Put("bar", "v2"); });
+  port::Thread writer2([&]() {
+    TEST_SYNC_POINT("ExternalSSTFileTest::WithUnorderedWrite:WaitWriteWAL");
+    ASSERT_OK(GenerateAndAddExternalFile(options, {{"bar", "v3"}}, -1,
+                                         true /* allow_global_seqno */));
+  });
+
+  // make sure `Put` and `GenerateAndAddExternalFile` are all called.
+  env_->SleepForMicroseconds(100000);
```

Comment:
Sleep almost always lead to flaky test. Can we get rid of it? Are we able to use the test from https://github.com/yiwu-arbug/rocksdb/commit/27399777f4280b291aa3e3c5733d61a3c62624e8 as is?

Reply:
No, cause after the fix, `Put` and `GenerateAndAddExternalFile` will block each other due to sync-point `NeedFlush` only can be called after `WaitForPendingWrites`. I tried multiple ways but only found sleeping can solve that.

Reply:
I think you can add a sync point in the beginning of `WaitForPendingWrites`(say P), and let the "DBImpl::WriteImpl:BeforeUnorderedWriteMemtable" wait for P instead. Remove the "ExternalSSTFileTest::WithUnorderedWrite:ContinueWriteMemtable" sync point. That way you don't need the sleep.

Reply:
if that, after removing the fix, it may still pass the test

Reply:
Removing the fix will remove `WaitForPendingWrites`, which makes the sync point waiting forever. Surely this is not an ideal unit test, but still better than a flaky one.

Reply:
oh, I misunderstood your meaning of `beginning`, seems reasonable.